Add an option to the Explorer component to navigate by tag

CVanmarcke opened the pull request:
This fork adds the option to make the Explorer build its content by the hierarchy of tags (and subtags) instead of the filetree. Every tag will be listed as a folder in the Explorer component, with its contents the pages which contain these tags and any subtags (also folders).

This can be toggled with the `content` option added to `Component.Explorer`: for example by adding `Component.Explorer({content: "tags",})),` to `quartz.layout.ts`.

Tag and subtag should be divided by a forward slash (example `tag/subtag`) to build the Explorer tree correctly.

This can be tested by adding `Component.Explorer({content: "tags",})),` to `quartz.layout.ts`.

Commit:
Added support to build explorer from tag instead of files.

To apply, add the "content" option with value "tags" to
Component.Explorer: eg Component.Explorer({content: "tags",})

Commit:
Updated the docs

Explains how to navigate by tag instead of by file hierarchy in the docs.

CVanmarcke commented:
Github provides automatic build errors, however running the code still works.

```
git clone https://github.com/CVanmarcke/quartz
cd quartz
npm i
npx quartz create
npx quartz build --serve
```

I am not very good at typescript (just learned it to add this commit), so I am unsure how to start testing for a fix without clogging up this commit.

EDIT: I figured out I can run the check myself with `npm run check`. I will spend time this evening fixing the errors.

aarnphm reviewed (commented):
Is there a case where you want the explorer to just explore tags?

my feeling is that the functionality of explorer is to see files not tags?

Btw you can already search for tags with <kbd>Ctrl+Shift+K</kbd>

CVanmarcke commented:
> Is there a case where you want the explorer to just explore tags?
> 
> my feeling is that the functionality of explorer is to see files not tags?
> 
> Btw you can already search for tags with Ctrl+Shift+K

Thanks for your reply.

I wanted to add this as currently, my whole knowledge database is imported from emacs org-roam. All the files of org-roam are in one single folder, instead of a folder with subfolders. 

All my files are tagged a certain way (eg. GI/rectum ; GI/duodenum ; liver/benign etcetera). By building the explorer tree with the tags, I still get a hierarchy but I don't have to put all my files in different folders.

Furthermore, this allows a file to exist in multiple "subfolders": for example, if I have a file with the tags "perineum/female" and "gynaecology/uterus", the file will appear in both "folders" using this method. If I only were to use a file hierarchy, the same file would only exist in one folder.

aarnphm commented:
Then I suggest that you should build a TagExplorer for this instead of coercing the explorer components.

explorer components is already a complicated codepath.

Also https://discord.com/channels/927628110009098281/1149197588084572161/1205044771870081064

But feel free to do this in your own vault.

Commit:
Revert "Updated the docs"

This reverts commit 0f1f4c3cda4a55cbe6b520b739a884b9d6f6ab2d.

Commit:
Revert "Added support to build explorer from tag instead of files."

This reverts commit ea5189e0b4db9cfb29fd42dec80f2e105e42d46c.

Commit:
Added TagExplorer component

This is a variant of Explorer to browse through a tag tree instead of
a file tree.

Commit:
Prettier

aarnphm commented:
from the discord message, we will pause all plugins/components addition for now.

There is a pr that will introduce plugins setup later. I will close it for now.

Feel free to do this for your own quartz.

CVanmarcke commented:
Thank you. I followed your advice and made TagExplorer in a different file. I managed to fix the errors and pass the tests (all those object types bring me back to my Java days!).

Ill wait till plugins are added and will reevaluate at a later time.

aarnphm commented:
Yes. I think when plugins is supported you can just write this yourself.

Then we will definitely have a page to feature all supported plugins.

Lorite commented:
Hi. I have a working version of the Tag Explorer in my fork. The code is very copy-pasty from the normal Explorer component. You can see it live here: https://alejandro.lorite.eu/

If you think it'd be a good idea to create a PR and merge it, please let me know. These are the commits:
- https://github.com/Lorite/lorite-personal-website-obsidian-quartz/commit/f41de39cce3f439fac8fb96fa81c2199f1c834ab
- https://github.com/Lorite/lorite-personal-website-obsidian-quartz/commit/71201229e9c33af3b91d49a4afadf8111f79501f
- https://github.com/Lorite/lorite-personal-website-obsidian-quartz/commit/f8ab5f330558d45f4fb454840db9ab838167ffcc
